File: ios/Sources/Data/ImageCleaner.swift
import Foundation
import ImageIO
import UniformTypeIdentifiers

/// Retire les métadonnées d'une image avant l'envoi.
///
/// Une photo prise au téléphone porte, par défaut, la position GPS de l'endroit
/// où elle a été prise, la date exacte et le modèle d'appareil. Sur un dossier
/// d'immigration, la localisation du domicile du client est une donnée sensible
/// qui fuirait sans qu'il le sache. On ré-encode donc l'image en ne gardant que
/// les pixels, jamais le dictionnaire EXIF ni GPS.
enum ImageCleaner {
    static func stripped(_ data: Data) -> Data {
        guard
            let source = CGImageSourceCreateWithData(data as CFData, nil),
            let type = CGImageSourceGetType(source),
            let image = CGImageSourceCreateImageAtIndex(source, 0, nil)
        else {
            // Un PDF ou un format que l'on ne sait pas relire part tel quel :
            // mieux vaut envoyer la pièce que la perdre. Les PDF ne portent pas
            // de GPS.
            return data
        }

        let out = NSMutableData()
        guard let dest = CGImageDestinationCreateWithData(out, type, 1, nil) else { return data }
        // On n'écrit aucune propriété : ni EXIF, ni GPS, ni TIFF. Que les pixels.
        CGImageDestinationAddImage(dest, image, [:] as CFDictionary)
        guard CGImageDestinationFinalize(dest) else { return data }
        return out as Data
    }
}

File: android/app/src/main/java/app/capmedia/visaflow/ImageCleaner.kt
package app.capmedia.visaflow

import android.graphics.BitmapFactory
import android.graphics.Bitmap
import java.io.ByteArrayOutputStream

/**
 * Retire les métadonnées d'une image avant l'envoi.
 *
 * Une photo prise au téléphone porte, par défaut, la position GPS de l'endroit
 * où elle a été prise. Sur un dossier d'immigration, la localisation du domicile
 * du client est sensible et fuirait sans qu'il le sache. On décode puis
 * recompresse l'image : le bitmap ne porte que les pixels, jamais l'EXIF.
 */
object ImageCleaner {
    fun stripped(data: ByteArray): ByteArray {
        // Un PDF ou un format non décodable part tel quel : mieux vaut envoyer
        // la pièce que la perdre, et un PDF ne porte pas de GPS.
        val bitmap: Bitmap = runCatching { BitmapFactory.decodeByteArray(data, 0, data.size) }
            .getOrNull() ?: return data
        return ByteArrayOutputStream().use { out ->
            // JPEG de bonne qualité, sans le moindre attribut EXIF recopié.
            bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out)
            bitmap.recycle()
            out.toByteArray()
        }
    }
}
